macOS/DuckDuckGo/WebExtensions/WebExtensionLoader.swift

```diff
@@ -18,39 +18,91 @@
 
 #if WEB_EXTENSIONS_ENABLED
 
+import CryptoKit
+
 @available(macOS 15.4, *)
 protocol WebExtensionLoading: AnyObject {
-
-    func loadWebExtensions(from paths: [String]) async -> [WKWebExtension]
-
+    @discardableResult
+    func loadWebExtension(path: String, into controller: WKWebExtensionController) async throws -> WKWebExtensionContext
+    func loadWebExtensions(from paths: [String], into controller: WKWebExtensionController) async -> [Result<WKWebExtensionContext, Error>]
+    func unloadExtension(at path: String, from controller: WKWebExtensionController) throws
 }
 
 @available(macOS 15.4, *)
 final class WebExtensionLoader: WebExtensionLoading {
 
-    private func loadWebExtension(path: String) async -> WKWebExtension? {
+    enum WebExtensionLoaderError: Error {
+        case failedToCreateURLFromPath(path: String)
+        case failedToFindContextForPath(path: String)
+    }
+
+    func loadWebExtension(path: String, into controller: WKWebExtensionController) async throws -> WKWebExtensionContext {
         guard let extensionURL = URL(string: path) else {
             assertionFailure("Failed to create URL from path: \(path)")
-            return nil
+            throw WebExtensionLoaderError.failedToCreateURLFromPath(path: path)
         }
-        let webExtension = try? await WKWebExtension(resourceBaseURL: extensionURL)
-        return webExtension
+
+        let webExtension = try await WKWebExtension(resourceBaseURL: extensionURL)
+        let context = makeContext(for: webExtension, at: path)
+        try await controller.load(context)
+        return context
     }
 
-    func loadWebExtensions(from paths: [String]) async -> [WKWebExtension] {
-        var result = [WKWebExtension]()
-        for webExtensionPath in paths {
-            guard let webExtension = await loadWebExtension(path: webExtensionPath) else {
-                assertionFailure("Failed to load the web extension: \(webExtensionPath)")
-                continue
+    func loadWebExtensions(from paths: [String], into controller: WKWebExtensionController) async -> [Result<WKWebExtensionContext, Error>] {
+        var result = [Result<WKWebExtensionContext, Error>]()
+        for path in paths {
+            do {
+                let context = try await loadWebExtension(path: path, into: controller)
+                result.append(.success(context))
+            } catch {
+                result.append(.failure(error))
             }
-
-            result.append(webExtension)
         }
 
         return result
     }
 
+    func unloadExtension(at path: String, from controller: WKWebExtensionController) throws {
+        let context = controller.extensionContexts.first {
+            $0.uniqueIdentifier == identifierHash(forPath: path)
+        }
+
+        guard let context else {
+            throw WebExtensionLoaderError.failedToFindContextForPath(path: path)
+        }
+
+        try controller.unload(context)
+    }
```

Comment:
The loader now also supports unloading extensions.

Reply:
Awesome! I tested it and it works great!